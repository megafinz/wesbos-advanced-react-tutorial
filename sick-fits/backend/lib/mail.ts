import { createTransport, getTestMessageUrl } from 'nodemailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: parseInt(process.env.MAIL_PORT),
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeEmail(text: string) {
  return `
    <div style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello there!</h2>
      <p>${text}</p>
    </div>`;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const result = await transport.sendMail({
    to,
    from: 'reset-password@sick-fits.com',
    subject: 'Your password reset token!',
    html: makeEmail(`Your password reset token is here!

      <a href="${process.env.FRONTEND_URL}/reset-password?token=${resetToken}">Click here to reset your password</a>
    `),
  });
  if (process.env.MAIL_USER.includes('ethereal.email')) {
    console.log(
      `✉️ Message sent! Preview it at ${getTestMessageUrl(result).toString()}`
    );
  }
}
