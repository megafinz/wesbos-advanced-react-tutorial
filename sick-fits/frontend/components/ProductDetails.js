import Head from 'next/head';
import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import DisplayError from './ErrorMessage';

const ProductStyles = styled.div`
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  max-width: var(--maxWidth);
  align-items: top;
  gap: 2rem;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
`;

const PRODUCT_DETAILS_QUERY = gql`
  query PRODUCT_DETAILS_QUERY($id: ID!) {
    Product(where: { id: $id }) {
      id
      name
      price
      description
      photo {
        altText
        image {
          publicUrlTransformed
        }
      }
    }
  }
`;

export default function ProductDetails({ id }) {
  const { data, loading, error } = useQuery(PRODUCT_DETAILS_QUERY, {
    variables: {
      id,
    },
  });
  if (loading) {
    return <p>Loadingâ€¦</p>;
  }
  if (error) {
    return <DisplayError error={error} />;
  }
  const { Product } = data;
  return (
    <ProductStyles>
      <Head>
        <title>Sick Fits | {Product.name}</title>
      </Head>
      <img
        src={Product.photo.image.publicUrlTransformed}
        alt={Product.photo.image.altText}
      />
      <div className="details">
        <h2>{Product.name}</h2>
        <p>{Product.description}</p>
      </div>
    </ProductStyles>
  );
}

ProductDetails.propTypes = {
  id: PropTypes.string,
};
