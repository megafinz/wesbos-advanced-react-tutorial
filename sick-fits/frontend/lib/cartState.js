import PropTypes from 'prop-types';
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

export function CartStateProvider({ children }) {
  const [cartOpen, setCartOpen] = useState(false);

  function openCart() {
    setCartOpen(true);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  return (
    <LocalStateProvider value={{ cartOpen, openCart, closeCart, toggleCart }}>
      {children}
    </LocalStateProvider>
  );
}

export function useCart() {
  return useContext(LocalStateContext);
}

CartStateProvider.propTypes = {
  children: PropTypes.any,
};
