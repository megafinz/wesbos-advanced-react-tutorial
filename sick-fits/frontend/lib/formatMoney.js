export default function formatMoney(amount) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
  };
  const formatter = new Intl.NumberFormat('en-us', options);
  return formatter.format(amount);
}
