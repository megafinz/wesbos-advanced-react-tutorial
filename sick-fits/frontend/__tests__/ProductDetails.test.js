import { MockedProvider } from '@apollo/react-testing';
import { render, screen } from '@testing-library/react';
import { fakeItem } from '../lib/testUtils';
import ProductDetails, {
  PRODUCT_DETAILS_QUERY,
} from '../components/ProductDetails';

const product = fakeItem();

const mocks = [
  {
    request: {
      query: PRODUCT_DETAILS_QUERY,
      variables: {
        id: '123',
      },
    },
    result: {
      data: {
        Product: product,
      },
    },
  },
  {
    request: {
      query: PRODUCT_DETAILS_QUERY,
      variables: {
        id: '1234',
      },
    },
    result: {
      errors: [{ message: 'Item not found!' }],
    },
  },
];

describe('<ProductDetails />', () => {
  it('renders with proper data', async () => {
    // Arrange / Act.
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <ProductDetails id="123" />
      </MockedProvider>
    );

    await screen.findByTestId('product-details-container');

    // Assert.
    expect(container).toMatchSnapshot();
  });

  it('errors out when item is not found', async () => {
    // Arrange / Act.
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <ProductDetails id="1234" />
      </MockedProvider>
    );

    await screen.findByTestId('graphql-error');

    // Assert.
    expect(container).toHaveTextContent('Shoot!');
    expect(container).toHaveTextContent('Item not found!');
  });
});
