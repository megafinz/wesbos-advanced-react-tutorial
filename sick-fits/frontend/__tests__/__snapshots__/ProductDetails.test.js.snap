// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> renders with proper data 1`] = `
<div>
  <div
    class="ProductDetails__ProductStyles-sc-199yvsf-0 bCtqoC"
    data-testid="product-details-container"
  >
    <img
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
